src/caffeine/collective_subroutines/co_max_s.f90

```diff
@@ -11,61 +11,67 @@
 
    interface
 
-     !! void c_co_max_no_result_image_int32(void* c_loc_a, int Nelem)
-     subroutine c_co_max_no_result_image_int32(c_loc_a, Nelem, c_loc_stat) bind(C)
+     !! void c_co_max_int32(void* c_loc_a, int Nelem)
+     subroutine c_co_max_int32(c_loc_a, Nelem, c_loc_stat, c_loc_result_image) bind(C)
        import c_ptr, c_size_t
-       type(c_ptr), value :: c_loc_a, c_loc_stat
+       type(c_ptr), value :: c_loc_a, c_loc_stat, c_loc_result_image
        integer(c_size_t), value :: Nelem
      end subroutine
 
-     subroutine c_co_max_no_result_image_int64(c_loc_a, Nelem, c_loc_stat) bind(C)
+     subroutine c_co_max_int64(c_loc_a, Nelem, c_loc_stat, c_loc_result_image) bind(C)
        import c_ptr, c_size_t
-       type(c_ptr), value :: c_loc_a, c_loc_stat
+       type(c_ptr), value :: c_loc_a, c_loc_stat, c_loc_result_image
        integer(c_size_t), value :: Nelem
      end subroutine
 
-     subroutine c_co_max_no_result_image_float(c_loc_a, Nelem, c_loc_stat) bind(C)
+     subroutine c_co_max_float(c_loc_a, Nelem, c_loc_stat, c_loc_result_image) bind(C)
        import c_ptr, c_size_t
-       type(c_ptr), value :: c_loc_a, c_loc_stat
+       type(c_ptr), value :: c_loc_a, c_loc_stat, c_loc_result_image
        integer(c_size_t), value :: Nelem
      end subroutine
 
-     subroutine c_co_max_no_result_image_double(c_loc_a, Nelem, c_loc_stat) bind(C)
+     subroutine c_co_max_double(c_loc_a, Nelem, c_loc_stat, c_loc_result_image) bind(C)
        import c_ptr, c_size_t
-       type(c_ptr), value :: c_loc_a, c_loc_stat
+       type(c_ptr), value :: c_loc_a, c_loc_stat, c_loc_result_image
        integer(c_size_t), value :: Nelem
      end subroutine
 
-    !subroutine c_co_max_no_result_image_char(c_loc_a, Nelem, c_loc_stat) bind(C)
+    !subroutine c_co_max_char(c_loc_a, Nelem, c_loc_stat) bind(C)
     !  import c_ptr, c_size_t
     !  type(c_ptr), value :: c_loc_a, c_loc_stat
     !  integer(c_size_t), value :: Nelem
     !end subroutine
 
    end interface
 
-   type(c_ptr) stat_ptr
+   type(c_ptr) stat_ptr, result_image_ptr
 
    if (present(stat)) then
      stat_ptr = c_loc(stat)
    else
      stat_ptr = c_null_ptr
    end if
 
+   if (present(result_image)) then
+     result_image_ptr = c_loc(result_image)
+   else
+     result_image_ptr = c_null_ptr
+   end if
```

Comment:
Why the level of indirection on `result_image`?

IIUC this forces the Fortran compiler to spill the result_image value to the stack in order to create a valid pointer, where it otherwise could remain in a register.

IIUC you could instead arrange to pass an `int` value argument to GASNet, where `0` means "all" (since Fortran images are 1-based).

Reply:
@bonachea   Wow. Really nice catch!  This is probably an example of copypasta (which isn't quite the write term but I loosely define it as spaghetti code that results from my tendency copy and paste working code at the beginning for writing new code).  I suspect that the only reason `result_image` is passed as a pointer is because it comes right after I `stat` in the argument list.  I think the arguments are listed in the order I implemented them so I probably copied whatever I was doing with `stat` and did the same with `result_image`!  I'll switch to just passing an `int` by value as you suggest.

Reply:
@bonachea I remember why I chose to make it a pointer.  There are several subtleties that I had to investigate in the standard before I could be sure that it would be safe to choose a special value indicate that no `result_image` was specified by the user.  For example, if a coarray `A` has multiple codimensions, then `this_image(A)` returns a vector indicating the image's position in a multidimensional index space wherein the range of each vector component can span positive and negative values.  Your comment prompted me to research this issue in the standard and it appears that image numbers are required to be positive so it should work to use a zero or negative `result_image` value to as a signal that no `result_image` was specified.

Reply:
@bonachea commit 343ebb3 switches to passing `result_image` by value.